### PR TITLE
internal libraries changes for new design apis

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:
This PR is sub PR of #740 and base PR of divided each component

This PR include many features like below 
- internal/io/ioutil for more easy ReadFile interface 
- new design internal/backoff.Backoff interface for more flexible backoff control
- v1 API design's internal/client libraries

**Please do not merge** until below PRs merge.
- PRXXX
- PRYYY

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kpango** [comment]
@rinx fixed

File: internal/client/v1/client/agent/grpc/option.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		if len(addr) != 0 {
+			c.addr = addr
+		}
+	}
+}
+
+// WithGRPCClientOption returns Option that sets options for gRPC.
+func WithGRPCClientOption(opts ...grpc.Option) Option {
+	return func(c *agentClient) {
+		if len(opts) != 0 {
```
Comment (hlts2) [inline internal/client/v1/client/agent/grpc/option.go:58]: Do you need fix based on coding-style ?
https://github.com/vdaas/vald/blob/master/docs/contributing/coding-style.md#functional-option

```go
if c. opts == nil {
        c. opts = opts
} else {
        c. opts = append(c. opts, opts...)
}

```

File: internal/client/v1/client/compressor/option.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+type Option func(c *client) error
+
+var (
+	defaultOpts = []Option{}
+)
+
+func WithAddr(addr string) Option {
+	return func(c *client) error {
+		c.addr = addr
```
Comment (hlts2) [inline internal/client/v1/client/compressor/option.go:32]: Do I need to check if the address is empty or not?

**Review by hlts2 — APPROVED** [review]
LGTM

File: hack/benchmark/assets/x1b/loader_test.go [hunk 3]
```diff
@@ -1,1288 +0,0 @@
-//
```
[the remark this answers is not in the document]
Comment (kpango) [reply hack/benchmark/assets/x1b/loader_test.go:1]: this code is auto-generated code, this code will be re-generate after #741 merged.

File: .github/workflows/dockers-gateway-backup-image.yml [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      - "pkg/gateway/internal/**"
+      - "dockers/base/Dockerfile"
+      - "dockers/gateway/backup/Dockerfile"
+      - "versions/GO_VERSION"
+  pull_request:
+    paths:
+      - "go.mod"
+      - "go.sum"
+      - "internal/**"
+      - "!internal/**/*_test.go"
```
Comment (kevindiu) [inline .github/workflows/dockers-gateway-backup-image.yml:30]:
```suggestion
      - "!internal/**/*_test.go"
      - "!internal/**/*_mock.go"
```

File: .github/workflows/dockers-gateway-lb-image.yml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    tags:
+       - '*.*.*'
+       - 'v*.*.*'
+       - '*.*.*-*'
+       - 'v*.*.*-*'
+    paths:
+      - 'go.mod'
+      - 'go.sum'
+      - 'internal/**'
+      - '!internal/**/*_test.go'
```
Comment (kevindiu) [inline .github/workflows/dockers-gateway-lb-image.yml:15]:
```suggestion
      - '!internal/**/*_test.go'
      - '!internal/**/*_mock.go'
```

File: .github/workflows/dockers-gateway-lb-image.yml [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      - 'pkg/gateway/internal/**'
+      - 'dockers/base/Dockerfile'
+      - 'dockers/gateway/lb/Dockerfile'
+      - 'versions/GO_VERSION'
+  pull_request:
+    paths:
+      - 'go.mod'
+      - 'go.sum'
+      - 'internal/**'
+      - '!internal/**/*_test.go'
```
Comment (kevindiu) [inline .github/workflows/dockers-gateway-lb-image.yml:30]:
```suggestion
      - '!internal/**/*_test.go'
      - '!internal/**/*_mock.go'
```

**Comment by vankichi** [comment]
@kpango it seems the test will be written overall ignoring current test, e.g. `internal/compress/`.
Do you have any plan to fix these tests?

**Comment by kpango** [comment]
I have no plan to fix tests.
which line is different from master?

**Comment by vankichi** [comment]
Sorry, it was my misunderstanding. 😓 
Please forget about it.

**Review by vankichi — APPROVED** [review]

**Comment by kpango** [comment]
@rinx @vankichi @kmrmt @hlts2 @kevindiu @datelier finally, all sub-branch are already merged !
I'm going to merge this PR, could you please re-review this branch before it ?

**Review by datelier — APPROVED** [review]

**Review by kmrmt — APPROVED** [review]

File: pkg/agent/core/ngt/handler/grpc/handler.go [hunk 7]
```diff
@@ -23,20 +23,28 @@ import (
...
 	"github.com/vdaas/vald/internal/observability/trace"
+	"github.com/vdaas/vald/internal/safety"
 	"github.com/vdaas/vald/pkg/agent/core/ngt/model"
 	"github.com/vdaas/vald/pkg/agent/core/ngt/service"
 )
 
-type Server agent.AgentServer
+type Server interface {
+	agent.AgentServer
+	vald.ValdServer
```
Comment (rinx) [inline pkg/agent/core/ngt/handler/grpc/handler.go:42]: it seems not to implement v1 APIs. is it okay? :thinking:

File: pkg/gateway/backup/README.md [hunk 8]
```diff
@@ -0,0 +1 @@
+# server sample
```
Comment (rinx) [inline pkg/gateway/backup/README.md:1]:
```suggestion
# vald backup gateway
```

File: pkg/agent/core/ngt/handler/grpc/handler.go (hunk 7 above)
Comment (kpango) [reply pkg/agent/core/ngt/handler/grpc/handler.go:42]: I'm looking into it, please wait a moment

**Comment by rinx** [comment]
If the agent-ngt code requires to implement v1 APIs, it is also needed to implement v1 APIs to meta, mangagers (backup, compressor, index), and discoverer.

 For the internal components (meta, managers, discoverer), I think it is okay to implement only v1 APIs.

**Comment by rinx** [comment]
Others LGTM

**Pull request merged**